subprojects/dependency-management/src/main/java/org/gradle/internal/resolve/ModuleVersionNotFoundException.java

```diff
@@ -128,29 +139,44 @@ private static void addLocations(TreeFormatter builder, Collection<String> locat
         }
         builder.node("Searched in the following locations");
         builder.startChildren();
-        for (String location : locations) {
-            builder.node(location);
-        }
+
+        locations.forEach(builder::node);
+
         builder.endChildren();
-        tryHintAboutArtifactSource(builder, locations);
     }
 
     // This method should ideally use more data to figure out if the message should be displayed
     // or not. In particular, the ivy patterns can make it difficult to find out if an Ivy artifact
     // source should be configured. At this stage, this information is lost, so we do a best effort
     // based on the file locations.
-    private static void tryHintAboutArtifactSource(TreeFormatter builder, Collection<String> locations) {
+    private void addPossibleResolution(Collection<String> locations) {
         if (locations.size() == 1) {
             String singleLocation = locations.iterator().next();
-            boolean isPom = singleLocation.endsWith(".pom");
-            boolean isIvy = singleLocation.contains("ivy-") && singleLocation.endsWith(".xml");
-            boolean isModule = singleLocation.endsWith(".module");
-            String format = isPom ? "Maven POM" : (isIvy ? "ivy.xml" : (isModule ? "Gradle module" : null));
+            String format = getFormatName(singleLocation);
             if (format != null) {
-                builder.node(
-                    String.format("If the artifact you are trying to retrieve can be found in the repository but without metadata in '%s' format, you need to adjust the 'metadataSources { ... }' of the repository declaration.", format)
-                );
+                resolutions = ImmutableList.of(String.format("If the artifact you are trying to retrieve can be found in the repository but without metadata in '%s' format, you need to adjust the 'metadataSources { ... }' of the repository declaration.", format));
             }
         }
     }
+
+    private static String getFormatName(String singleLocation) {
+        boolean isPom = singleLocation.endsWith(".pom");
+        boolean isIvy = singleLocation.contains("ivy-") && singleLocation.endsWith(".xml");
+        boolean isModule = singleLocation.endsWith(".module");
+        return isPom ? "Maven POM" : (isIvy ? "ivy.xml" : (isModule ? "Gradle module" : null));
+    }
+
+    @Override
+    public List<String> getResolutions() {
+        return resolutions;
+    }
+
+    protected ModuleVersionResolveException createCopy() {
+        try {
+            String message = getMessage();
+            return getClass().getConstructor(ComponentSelector.class, Factory.class, Collection.class).newInstance(selector, (Factory<String>) () -> message, resolutions);
```

Comment:
❔ Why is this using reflection instead of: `return new ModuleVersionNotFoundException(getSelector(), this::getMessage, resolutions);`?